apps/src/dance/Dance.js

```diff
@@ -148,6 +148,17 @@ Dance.prototype.init = function(config) {
 
   this.initSongsPromise = this.initSongs(config);
 
+  // As students add new characters to their programs,
+  // load sprites asynchronously. Previously, we waited to load
+  // sprites once students hit the run button.
+  // This is in place to support iOS, which will not play audio
+  // if started asynchronously after fetching sprites.
```

Comment:
Just to double check. If a student uses a new sprite and then immediately clicks the run button before the download has completed, will this still work? (thinking esp. for slow internet connections)

Reply:
I believe it would "work" as in their program would still run, but it's possible that the audio might still have issues. We still check to ensure that sprites are loaded when a student clicks run, but if during that check we find that there are sprites still to be loaded, I think we'd see the existing audio issues:

https://github.com/code-dot-org/code-dot-org/blob/687ce11b4d1b1ac1f4c5df63748824d6c6d68c3b/apps/src/dance/Dance.js#L577-L592

My thought was that this would still be an improvement for most students, but definitely see this as a limitation of this approach. I could do more testing with a throttled connection though to see what happens, although it appears Safari doesn't allow you to throttle your connection easily in devtools, and Safari is how I'm debugging my iPhone locally. I could check how the download times change on sprites between different connection speeds on my laptop perhaps?

Reply:
FWIW, since we cache these assets between runs, I think students on slow connections might have the audio not playing issue on their first run (if we have to make requests to get sprites), but wouldn't encounter the issue on subsequent runs (if they're using the same assets).

Reply:
Ok. That definitely sounds like an improvement to me 👍

Reply:
It seems, ideally, that the ability to play sound would be independent of whether we needed to load sprites or not.  Are we delaying the playback until after they are loaded so that we know that we can keep the visuals synchronized to the music?  Or for architectural simplicity?

Reply:
To be honest, I don't know -- being new to the space, I operated under the assumption that we wait to load sprites before playing sound for some (good) reason, maybe for synchronization like you said? Do you think knowing the "why" might change the approach proposed here?

Reply:
Unfortunately I'm not familiar with the original design decisions here.  But I'm sure we do want the animation to be in sync with the music, and if we're not ready to play the animation yet because the images aren't loaded, then it makes sense that we'll need to delay playing the music.